ipam/pool_test.go

```diff
@@ -0,0 +1,1007 @@
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["10.0.0.0/16"] = &addressPool{
+						Id:       "10.0.0.0/16",
+						Priority: 1,
+					}
+					as.Pools["10.1.0.0/16"] = &addressPool{
+						Id:       "10.1.0.0/16",
+						Priority: 2,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("10.1.0.0/16"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When addressSpace has pools with different addresses", func() {
+				It("Should return the pool with the highest number of addresses", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["10.0.0.0/16"] = &addressPool{
+						Id:        "10.0.0.0/16",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["10.1.0.0/16"] = &addressPool{
+						Id:        "10.1.0.0/16",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["10.1.0.0/16"].Addresses["10.1.0.1/16"] = &addressRecord{}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("10.1.0.0/16"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+		})
+
+		Describe("Test releasePool", func() {
+			Context("When pool not found", func() {
+				It("Should raise an error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					err := as.releasePool("10.0.0.0/16")
+					Expect(err).To(Equal(errAddressPoolNotFound))
+				})
+			})
+
+			Context("When pool is not in use", func() {
+				It("Should raise an error", func() {
+					poolId := "10.0.0.0/16"
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools[poolId] = &addressPool{
+						RefCount: 0,
+					}
+					err := as.releasePool("10.0.0.0/16")
+					Expect(err).To(Equal(errAddressPoolNotInUse))
+					Expect(as.Pools[poolId]).NotTo(BeNil())
+				})
+			})
+
+			Context("When pool's epoch is less than the space's epoch and pool is never in use", func() {
+				It("Should release the pool ", func() {
+					poolId := "10.0.0.0/16"
+					as := &addressSpace{
+						epoch: 2,
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools[poolId] = &addressPool{
+						epoch:    1,
+						RefCount: 1,
+					}
+					err := as.releasePool("10.0.0.0/16")
+					Expect(err).NotTo(HaveOccurred())
+					Expect(as.Pools[poolId]).To(BeNil())
+				})
+			})
+
+			Context("When the epoch of pool is equal to the epoch of addressSpace", func() {
+				It("Should not delete the pool", func() {
+					poolId := "10.0.0.0/16"
+					as := &addressSpace{
+						epoch: 1,
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools[poolId] = &addressPool{
+						epoch:    1,
+						RefCount: 1,
+					}
+					err := as.releasePool("10.0.0.0/16")
+					Expect(err).NotTo(HaveOccurred())
+					Expect(as.Pools[poolId]).NotTo(BeNil())
+				})
+			})
+
+			Context("When pool is still in use", func() {
```

Comment:
whats the difference between this test and this one -> Context("When pool's epoch is less than the space's epoch and pool is never in use"

Reply:
so is releasing pool not happening because refcount is 2. If that is 1, release pool would have succeeded?

Reply:
Yes, in releasePool, we first `ap.RefCount--` and then check if pool is in use, so when the refcount is 2, it will be set to 1 and still in use, but when the refcount is 1, it will be set to 0 and not still in use.